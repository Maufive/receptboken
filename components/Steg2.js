import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyledForm } from "./styles/FormStyles";
import EditIcon from "../svg/edit.svg";
import AddIcon from "../svg/add.svg";
import IngredientIcon from "../svg/groceries.svg";
import TrashIcon from "../svg/trash.svg";
import { List, ListItemDiv } from "./styles/Steg1Styles";
import { Button, DisabledButton } from "./styles/Button";

class Steg2 extends Component {
	state = {
		arr: [],
		string: "",
		step: 1,
		editing: false,
		editIndex: null
	};

	componentDidMount() {
		this.setState({
			arr: this.props.description
		});
	}

	addToArray = e => {
		const { string, step } = this.state;
		e.preventDefault();

		if (string.length >= 5) {
			const steg = `Steg ${step}: `;
			const completeString = steg.concat(" ", string);

			this.setState(prevState => ({
				arr: [...prevState.arr, completeString],
				step: step + 1,
				string: ""
			}));
		} else {
			this.props.setMessage("info", "Du måste skriva minst 5 bokstäver");
		}
	};

	deleteItem = e => {
		// 1. Gör en kopia av arrayen i state
		let arr = [...this.state.arr];
		// 2. Hitta rätt item att ta bort
		let index = arr.indexOf(e);
		// 3.
		if (index !== -1) {
			arr.splice(index, 1);
			this.setState({ arr });
		}
	};

	editItem = currentItem => {
		// 1. Gör en kopia av state
		let items = [...this.state.arr];
		// 2. Gör en kopia av det item jag vill ändra
		let itemIndex = items.indexOf(currentItem);
		// 3. Ändra item
		this.setState({
			string: items[itemIndex],
			editing: true,
			editIndex: itemIndex
		});
		this.props.setMessage("success", "Du kan nu ändra!");
	};

	saveEdit = e => {
		e.preventDefault();
		const { string } = this.state;
		const index = this.state.editIndex;
		const steg = `Steg ${index + 1}: `;
		const completeString = steg.concat(" ", string);
		// kolla så det finns något i inputen
		if (string.length >= 2) {
			// ta en kopia på arrayen i state för att modifiera
			const newArr = [...this.state.arr];
			// ta den redigerade strängen och tryck in den på det index i arrayen som jag hämtat från editItem funktionen
			newArr[index] = completeString;
			// spara.
			this.setState({
				arr: newArr,
				string: "",
				editing: false,
				message: null
			});
		} else {
			this.props.setMessage("info", "Du måste skriva in minst 5 bokstäver!");
		}
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	nextStep = e => {
		e.preventDefault();
		this.props.saveDescription(this.state.arr);
	};

	// prevStep = e => {
	// 	e.preventDefault();
	// 	this.props.previousStep();
	// };

	render() {
		const { steg, previousStep } = this.props;
		return (
			<StyledForm>
				<h3 style={{ display: "flex", justifyContent: "space-between" }}>
					<span>Ingedienser</span>
					<span>Steg {steg} av 3</span>
				</h3>
				<div>
					<textarea
						name="string"
						required
						value={this.state.string}
						onChange={this.saveToState}
					/>
					<label htmlFor="string">Fyll i Steg {this.state.step}</label>
				</div>
				<div>
					{this.state.editing ? (
						<Button onClick={e => this.saveEdit(e, this.state.editIndex)}>
							<EditIcon />
							Ändra
						</Button>
					) : (
						<Button fullWidth onClick={this.addToArray}>
							<AddIcon />
							Lägg till
						</Button>
					)}
				</div>
				<List>
					{this.state.arr.map(item => (
						<li key={item}>
							<ListItemDiv>
								<p>{item}</p>
								<div>
									<TrashIcon onClick={() => this.deleteItem(item)} />
									<EditIcon onClick={() => this.editItem(item)} />
								</div>
							</ListItemDiv>
						</li>
					))}
				</List>
				<div style={{ display: "flex" }}>
					<Button fullWidth onClick={previousStep}>
						← Föregående
					</Button>
					{this.state.arr.length <= 1 && (
						<DisabledButton fullWidth>Nästa steg →</DisabledButton>
					)}
					{this.state.arr.length >= 2 && (
						<Button primary fullWidth onClick={this.nextStep}>
							Nästa steg →
						</Button>
					)}
				</div>
			</StyledForm>
		);
	}
}

export default Steg2;

Steg2.propTypes = {
	toNextStep: PropTypes.func,
	previousStep: PropTypes.func,
	steg: PropTypes.number,
	description: PropTypes.array,
	saveDescription: PropTypes.func
};
