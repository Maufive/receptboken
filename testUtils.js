export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeShoppinglist = [
	{
		_id: "34567876543452176",
		title: "My Shoppinglist",
		created: "12/3 2019",
		author: "61386712376576d7q6ewd65",
		list: [
			{
				input: "Lök",
				numberOfUnits: "5",
				units: "dl"
			},
			{
				input: "Tomater",
				numberOfUnits: "10",
				units: "st"
			}
		]
	}
];
