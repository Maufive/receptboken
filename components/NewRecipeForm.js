import React, { Component } from "react";
import { Header, RecipeForm } from "./styles/Steg1Styles";
import Steg1 from "./Steg1";
import Steg2 from "./Steg2";
import Steg3 from "./Steg3";

class NewRecipeForm extends Component {
	state = {
		title: "",
		description: [],
		tags: [],
		timeRequired: 0,
		photo: "",
		ingredients: [],
		servings: 2,
		step: 1
	};

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	previousStep = () => {
		this.setState({ step: this.state.step - 1 });
	};

	saveIngredients = (arr, servings) => {
		this.setState({
			ingredients: arr,
			servings,
			step: this.state.step + 1
		});
	};

	saveDescription = arr => {
		this.setState({
			description: arr,
			step: this.state.step + 1
		});
	};

	saveDetails = (photo, time, tags) => {
		this.setState({
			photo,
			timeRequired: time,
			tags,
			step: this.state.step + 1
		});
	};

	render() {
		const step = this.state.step;
		let ActiveComponent;
		// För varje steg i processen att skapa ett nytt recept så renderar jag en specifik component.
		// Syftet är att dela upp en lång process att ladda upp ett recept i mindre bitar för att göra det
		// lättare för användare.
		if (step === 1) {
			ActiveComponent = (
				<Steg1
					saveToState={this.saveToState}
					saveIngredients={this.saveIngredients}
					ingredients={this.state.ingredients}
					servings={this.state.servings}
				/>
			);
		} else if (step === 2) {
			ActiveComponent = (
				<Steg2
					saveDescription={this.saveDescription}
					steg={step}
					previousStep={this.previousStep}
					description={this.state.description}
				/>
			);
		} else if (step === 3) {
			ActiveComponent = (
				<Steg3
					steg={step}
					previousStep={this.previousStep}
					saveDetails={this.saveDetails}
				/>
			);
		} else {
			ActiveComponent = <h2>Finished!</h2>;
			// To-do: Skriv en funktion för en POST-method att passa ner till steg 3. Spara Allt som finns i state här
			// till databasen.
			// Gör en LOADING komponent som snurrar medans man väntar på att databasen ska spara receptet
			// Gör en Page för ett recept där jag kan displaya informationen
			// Redirecta användaren när man lyckats spara ett recept till databasen till det receptet

		}

		return (
			<div>
				<Header>
					<input
						type="text"
						value={this.state.title}
						onChange={this.saveToState}
						name="title"
						required
						placeholder="Nytt recept..."
					/>
					<i onClick={this.editTitle} className="icofont-ui-edit" />
				</Header>
				<RecipeForm>{ActiveComponent}</RecipeForm>
			</div>
		);
	}
}

export default NewRecipeForm;
