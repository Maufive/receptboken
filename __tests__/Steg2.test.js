import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import Steg2 from "../components/Steg2";
import NewRecipeForm from "../components/NewRecipeForm";
import { fakeUser } from "../testUtils";

describe("<Steg2 />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<Steg2 />)));

	it("renders without error and matches snapshot", () => {
		expect(wrapper.find("FormStyles__StyledForm").exists()).toBe(true);
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	describe("Textarea input", () => {
		it("renders without error", () => {
			expect(wrapper.find("#string").length).toBe(1);
		});

		it("handles the onChange event", () => {
			wrapper.find("#string").simulate("change", {
				target: { name: "string", value: "Detta är ett test" }
			});

			expect(wrapper.state().string).toEqual("Detta är ett test");
		});
	});

	describe("Previous / Next buttons", () => {
		it("return button decrements the Step-variable", () => {
			// WORK IN PROGRESS
			// const Parent = shallow(<NewRecipeForm user={fakeUser} />);
			// const mockFn = {
			// 	preventDefault: jest.fn()
			// };
			// Parent.setState({
			// 	step: 2
			// });
			// Parent.find("#button-previous").simulate("click", mockFn);
			// // expect(mockFn.preventDefault).toHaveBeenCalledTimes(1);
			// console.log(Parent.state());
		});
	});
});
