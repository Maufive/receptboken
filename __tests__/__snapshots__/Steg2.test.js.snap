// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steg2 /> renders without error and matches snapshot 1`] = `
<FormStyles__StyledForm>
  <h3>
    <span>
      Gör så här:
    </span>
    <span>
      Steg 
       av 3
    </span>
  </h3>
  <div>
    <textarea
      id="string"
      name="string"
      onChange={[Function]}
      required={true}
      value=""
    />
    <label
      htmlFor="string"
    >
      Fyll i beskrivning
    </label>
  </div>
  <div>
    <Button
      fullWidth={true}
      onClick={[Function]}
    >
      <AddBoldIcon
        viewBox="0 0 42 42"
        xmlns="http://www.w3.org/2000/svg"
      />
      Lägg till
    </Button>
  </div>
  <Steg1Styles__List />
  <div
    style={
      Object {
        "display": "flex",
      }
    }
  >
    <Button
      fullWidth={true}
      id="button-previous"
    >
      ← Föregående
    </Button>
  </div>
</FormStyles__StyledForm>
`;
